### PR TITLE
show audio format in status bar

### PR DESCRIPTION
**Pull request by adipose**
Added some code to translate audio formats to strings.  New function `GetAudioStreamInfo` based on `IsValidAudioStream`, to return the format string.

Convenience function to add strings with delimiters, can be reused more probably for status bar.

I don't see a ton of value in what else `OpenSetupStatusBar` is doing.  It loops on filters twice, one time apparently to grab some limited video info, and the second time to decide if the audio is Mono or not (even 7.1 will show as "stereo").  I'm pretty sure that could be gotten more concisely with `GetAudioStreamInfo`, but do we even need to do that?

If we want to keep all the behavior, I could probably shorten this routine a lot.

### CODE REVIEW

File: src/mpc-hc/MainFrm.cpp [hunk 1]
```diff
@@ -9264,7 +9264,7 @@ void CMainFrame::OnPlayShadersPresets(UINT nID)
     }
 }
 
-bool CMainFrame::IsValidAudioStream(int i) {
+bool CMainFrame::GetAudioStreamInfo(int i, bool loadFormat, CStringW& audioFormat) {
```
Comment (clsid2) [inline src/mpc-hc/MainFrm.cpp]: Maybe "extractFormatStr" is a clearer name for the bool?

Comment (adipose) [reply src/mpc-hc/MainFrm.cpp]: Oh, woops, I meant to fix that.

File: src/DSUtil/DSUtil.cpp [hunk 2]
```diff
@@ -2031,3 +2031,30 @@ CStringW ForceTrailingSlash(CStringW folder) {
     }
     return folder;
 }
+
+CStringW GetShortAudioNameFromMediaType(AM_MEDIA_TYPE* pmt) {
+    if (pmt->subtype == MEDIASUBTYPE_MPEG_ADTS_AAC) {
+        return L"ADTS";
```
Comment (clsid2) [inline src/DSUtil/DSUtil.cpp]: ADTS might be confusing for some people. Probably better to use just `AAC`.

File: src/DSUtil/DSUtil.cpp [hunk 3]
```diff
@@ -2031,3 +2031,30 @@ CStringW ForceTrailingSlash(CStringW folder) {
     }
     return folder;
 }
+
+CStringW GetShortAudioNameFromMediaType(AM_MEDIA_TYPE* pmt) {
+    if (pmt->subtype == MEDIASUBTYPE_MPEG_ADTS_AAC) {
+        return L"ADTS";
+    } else if (pmt->subtype == MEDIASUBTYPE_DTS_HD) {
+        return L"DTS-HD";
+    } else if (pmt->subtype == MEDIASUBTYPE_MPEG_HEAAC) {
```
Comment (clsid2) [inline src/DSUtil/DSUtil.cpp:2068]: Maybe just `AAC`.

File: src/DSUtil/DSUtil.cpp [hunk 4]
```diff
@@ -2031,3 +2031,30 @@ CStringW ForceTrailingSlash(CStringW folder) {
...
+        return L"HE-AAC";
+    } else if (pmt->subtype == MEDIASUBTYPE_MPEG_LOAS) {
+        return L"LOAS";
+    } else if (pmt->subtype == MEDIASUBTYPE_DOLBY_DDPLUS) {
+        return L"DD+";
+    } else if (pmt->subtype == MEDIASUBTYPE_DOLBY_AC3 || pmt->subtype == MEDIASUBTYPE_DOLBY_AC3_SPDIF || pmt->subtype == MEDIASUBTYPE_DVM || pmt->subtype == MEDIASUBTYPE_RAW_SPORT || pmt->subtype == MEDIASUBTYPE_SPDIF_TAG_241h) {
+        return L"AC3";
+    } else if (pmt->subtype == MEDIASUBTYPE_IEEE_FLOAT) {
+        return L"IEEE-FLOAT";
```
Comment (clsid2) [inline src/DSUtil/DSUtil.cpp]: `PCM`

File: src/DSUtil/DSUtil.cpp [hunk 5]
```diff
@@ -2031,3 +2031,30 @@ CStringW ForceTrailingSlash(CStringW folder) {
...
+    } else if (pmt->subtype == MEDIASUBTYPE_DTS || pmt->subtype == MEDIASUBTYPE_DTS2) {
+        return L"DTS";
+    } else if (pmt->subtype == MEDIASUBTYPE_DVD_LPCM_AUDIO) {
+        return L"LPCM";
+    } else if (pmt->subtype == MEDIASUBTYPE_MPEG1AudioPayload || pmt->subtype == MEDIASUBTYPE_MPEG1Packet || pmt->subtype == MEDIASUBTYPE_MPEG1Payload) { //are these all actually possible?
+        return L"MP1";
+    } else if (pmt->subtype == MEDIASUBTYPE_MPEG2_AUDIO) {
+        return L"MP2";
+    }
+    return L"";
```
Comment (clsid2) [inline src/DSUtil/DSUtil.cpp:2108]: Other common ones to add are:
FLAC, Opus, Vorbis

Lots more can be added in future.

**Comment by clsid2** [comment]
The code in `OpenSetupStatusBar` is used for the audio icon next to the timer in statusbar. But it could be extended to actually store the channel count in string format to be displayed together with the audio format.

Add convenience function to convert the channel count plus channel mask into string like: 2.0 / 5.1 / 7.1

File: src/mpc-hc/MainFrm.cpp [hunk 6]
```diff
@@ -9264,23 +9264,37 @@ void CMainFrame::OnPlayShadersPresets(UINT nID)
...
+            }
             return true;
         }
     }
     return false;
 }
 
+bool CMainFrame::IsValidAudioStream(int i) {
+    CStringW discard;
+    return GetAudioStreamInfo(i, false, discard);
```
Comment (clsid2) [inline src/mpc-hc/MainFrm.cpp]: two returns

File: src/mpc-hc/MainFrm.cpp (hunk 1 above)
Comment (clsid2) [reply src/mpc-hc/MainFrm.cpp]: currently only changed in header

**Comment by adipose** [comment]
> The code in `OpenSetupStatusBar` is used for the audio icon next to the timer in statusbar. But it could be extended to actually store the channel count in string format to be displayed together with the audio format.
> 
> Add convenience function to convert the channel count plus channel mask into string like: 2.0 / 5.1 / 7.1

Right, but I wonder, do we care about showing mono vs. stereo, or maybe we should have an icon for 2.0/5.1/7.1?

**Comment by adipose** [comment]
Oh, I also need to see about DVD and if audio lookup still works.

**Comment by clsid2** [comment]
That tiny icon is mainly to indicate if there is audio or not. Play a file without audio and see. Volume/mute still work if there is no audio.

Other common audio formats to add:
TrueHD/MLP, DTS-HD

File: src/DSUtil/DSUtil.cpp [hunk 7]
```diff
@@ -2032,7 +2032,25 @@ CStringW ForceTrailingSlash(CStringW folder) {
     return folder;
 }
 
+CStringW GetChannelStrFromMediaType(AM_MEDIA_TYPE* pmt) {
+    if (pmt && pmt->majortype == MEDIATYPE_Audio && pmt->formattype == FORMAT_WaveFormatEx) {
+        CStringW ret;
+        switch (((WAVEFORMATEX*)pmt->pbFormat)->nChannels) {
```
Comment (clsid2) [inline src/DSUtil/DSUtil.cpp]: int nChannels = ...

**Comment by clsid2** [comment]
For super uncommon audio formats the "FourCC" could be extracted if mediatype has the generic format:
`{0000ABCD-0000-0010-8000-00AA00389B71}`
In example it is ABCD.

**Comment by bibsp** [comment]
> That tiny icon is mainly to indicate if there is audio or not.

My idea. If there is no audio track then indicate it by disabling the spekar button on toolbar. Or maybe add a cross on it. Then you can remove the status bar icon.

Edit: For Dolby digital plus please show E-AC3 instead of DD+.

**Comment by clsid2** [comment]
Not possible. The speaker button only has enabled and muted state, and muted is not same as no audio.

**Comment by adipose** [comment]
Added basic support for audio for DVD.  Could be improved by breaking up the channel formatter.  Right now it just says "2ch" etc.

I don't think DVD is showing video codec?  Could be done in same area.

**Comment by clsid2** [comment]
I have made several modifications. It is now also combined with gathering the language info.

Stuff yet to do:
- Handle track change when done though splitter tray icon or context menu
- Track change with audio switcher disabled
- DVD audio language is not set yet at initial setup.

**Comment by clsid2** [comment]
Track change in splitter is a bit problematic. Afaik there is no event other than the generic `EC_PAUSED` graph event.

Checking selected track each time this event occurs would suck a bit, since it happens often when for example frame stepping.
`m_audioTrackCount` could be used to skip if there is just one stream.
Similar vars could be added for amount of embedded/external subs. But generally there are multiple of those.
For subs it can be skipped if external sub is active one, since splitter selection is irrelevant in that case.
Updating can of course also be skipped if statusbar options are disabled.

**Comment by dbojan** [comment]
Just a note. Haven't seen it written anywhere, but if you want to enable codec name display in status bar, download newer version of mpc-hc, go to 'view\options\user interface', and enable 'audio info', and other options like 'video info' at the bottom of the screen.

You can also see codec and bit rate, if you enable 'view\show statistics', or right click on main screen, select 'audio track' or 'video track' ...

**Pull request merged**